RealmSwift/Logger.swift

```diff
@@ -0,0 +1,275 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2024 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import Realm
+import Realm.Private
+
+/**
+ `Logger` is used for creating your own custom logging logic.
+
+ You can define your own logger creating an instance of `Logger` and define the log function which will be
+ invoked whenever there is a log message.
+
+ ```swift
+ let logger = Logger(level: .all, category: Category.realm) { level, message in
+    print("Realm Log - \(category.toString())-\(level): \(message)")
+ }
+ ```
+
+ Set this custom logger as you default logger using `Logger.shared`.
+
+ ```swift
+    Logger.shared = inMemoryLogger
+ ```
+
+ - note: By default default log threshold level is `.info`, and logging strings are output to Apple System Logger.
+*/
+public typealias Logger = RLMLogger
+extension Logger {
+    /**
+     Log a message to the supplied level.
+
+     ```swift
+     let logger = Logger(level: .info, logFunction: { level, message in
+         print("Realm Log - \(level): \(message)")
+     })
+     logger.log(level: .info, message: "Info DB: Database opened succesfully")
+     ```
+
+     - parameter level: The log level for the message.
+     - parameter category: The log category for the message.
+     - parameter message: The message to log.
+     */
+    internal func log(level: LogLevel, category: LogCategory = Category.realm, message: String) {
+        self.log(with: level, category: category.toString(), message: message)
+    }
+
+    /**
+     Creates a logger with the associated log level, category and the logic function to define your own logging logic.
+
+     ```swift
+     let logger = Logger(level: .info, category: Category.All, logFunction: { level, category, message in
+         print("\(category.toString()) - \(level): \(message)")
+     })
+     logger.log(level: .info, category: Category.realm, message: "Info DB: Database opened succesfully")
```

Comment:
This is not a use-case we want to support. There's no reason for users to be sending log messages through our mechanism for reporting logs to them.